### ISSUE TITLE
Fix and move the filter functionality in ScheduleFragmentViewModel

### ISSUE THREAD

**Issue by dr0pdb** [fossasia/open-event-droidgen#2066]
**Actual Behaviour**

The filter functionality of ScheduleFragment is implemented inside the Fragment. It isn't working properly as well.

**Expected Behaviour**

It should be fixed and moved inside the ViewModel

**Would you like to work on the issue?**

Yes

**Comment by divyabaid16** [comment]
I can't find ViewModel class. Do I have to implement it ?

**Comment by dr0pdb** [comment]
Actually i am already working on this issue as i have stated on the issue itself.

**Comment by simarsingh24** [comment]
@srv-twry are you still on it? If not can i take over it?

**Comment by dr0pdb** [comment]
Yes, i intend to work on it. Actually i have a PR on another issue in which i have a doubt. I have the same doubt in this issue also. Once that doubt is cleared and that PR is merged, i will fix this one also.
Thanks

**Comment by dr0pdb** [comment]
@simarsingh24 I am already very occupied with other issues hence if you want, you may go ahead with this. I think you're also occupied with implementing the Event invoices and Discount codes, hence if you want we may hand it over to @mayank8318 who could do it a bit quicker.

**Comment by simarsingh24** [comment]
@srv-twry sure! i will start working on it :)

**Comment by simarsingh24** [comment]
@srv-twry the core logic for the filter is written inside DayScheduleFragment and the latter doesn't have a ViewModel what should be done?
`((DayScheduleFragment) adapter.getItem(position)).filter(selectedTracks);`

**Comment by dr0pdb** [comment]
Oh, i realised now that there are two types of filtering in the ```ScheduleFragment```. The filtering mentioned in this issue is actually in the method ```filterSchedule```. 

As far as the filtering you're talking about, we'll have to wait for the ViewModel to be created. It has a PR but it hasn't been updated for some time. I'll myself create it if it isn't updated in the next couple of days.

**Comment by simarsingh24** [comment]
@mayank8318 I am a bit occupied right now as I have an interview coming and also other active issues, you can go ahead and work on it if you want, (I will finish it if you are also busy). Do drop a comment below and let us know if you are going to work on it

**Comment by sejalgoyal** [comment]
@simarsingh24 I will start working on it.

**Comment by simarsingh24** [comment]
@sejalgoyal please go ahead !

**Comment by mayank8318** [comment]
@iamareebjamal If I am not wrong the current implementation is as follows:
1) Query all the tracks for the day and sort according to the sort order specified
2) Make a list of `selectedTracks` as per the tracks selected in the filter dialog by the user( by default we assume everything is selected)
3) Execute the String Search Query on this `selectedTracks` to narrow down on a final list which is displayed to the user.
Please correct me if I am wrong

**Comment by iamareebjamal** [comment]
Yes

**Comment by mayank8318** [comment]
@iamareebjamal This seems to be a pretty good model. What improvement would you like to see here?

**Comment by iamareebjamal** [comment]
Lots of memory leaks and working with arrays. The implementation has problems, not the model. Neither it is reactive and present in ViewModel making it tightly coupled with UI

**Comment by mayank8318** [comment]
@iamareebjamal Ok what about this type of implementation ->
We only keep one arraylist in the UI. This will be displayed to the user. All the manipulations will happen in the View Model and this arraylist will be returned to the UI and the UI shall update accordingly. The UI won't know how the array is being manipulated, it will only display it and call for view model methods for manipulation when the user chooses to filter etc.

**Comment by iamareebjamal** [comment]
yes
